File: src/LeetCode380.cpp
class RandomizedSet {
private:

    unordered_map<int, int> map;
    vector<int> vec;

public:

    RandomizedSet() {

    }


    bool insert(int val) {
        if (map.find(val) != map.end()) {
            return false;
        }

        map[val] = vec.size();
        vec.push_back(val);
        return true;
    }


    bool remove(int val) {
        if (map.find(val) == map.end()) {
            return false;
        }

        int index = map[val];
        vec[index] = vec[vec.size() - 1];
        map[vec[vec.size() - 1]] = index;
        map.erase(val);
        // vec.erase(vec.begin() + vec.size() - 1);
        vec.pop_back();
        return true;
    }


    int getRandom() {
        int index = rand() % vec.size();
        return vec[index];
    }
};

File: src/LeetCode380.java
class RandomizedSet {

    Map<Integer, Integer> map; // val, index
    List<Integer> list;

    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
    }

    public boolean insert(int val) {
        if (map.containsKey(val)) {
            return false;
        }

        map.put(val, list.size());
        list.add(val);
        return true;
    }

    public boolean remove(int val) {
        if (!map.containsKey(val)) {
            return false;
        }
        int index = map.get(val);
        list.set(index, list.get(list.size() - 1));
        map.put(list.get(list.size() - 1), index);
        map.remove(val);
        list.remove(list.size() - 1);
        return true;
    }

    public int getRandom() {
        Random rand = new Random();
        int index = rand.nextInt(list.size());
        return list.get(index);
    }
}
